test-suite/squidconf/tls-certs

```diff
@@ -0,0 +1,256 @@
+## Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+#
+# The following tests check the sparse matrix of (valid) configs:
```

Comment:
Does `sparse` imply that some valid cases could be included but where omitted? If yes, that makes the reader wonder which cases were omitted and why. If you were able to enumerate all of the cases generated by the patterns listed below, then I suggest documenting that accomplishment:

```suggestion
# The following port configurations check all permutations of these configuration
# parameters. The unsupported and invalid ones are currently commented out.
# TODO: Check each negative (currently commented out) test case in isolation.
```

Reply:
This is one of the change requests that was resolved unaddressed.

Reply:
"sparse matrix" is a grid of permutatinos where most combinations are nil values. In specific to this, they are invalid combinations of things. eg key stored in separate file AND no key present,  key stored in cert file AND absent, chain stored in cert file AND absent.

Reply:
"just the key" should be the (A5 x B1 x C1-5) set you have there as second line in your 45 calculation.

The extra 5 lines are the (A5 x B2 x C1-5) set you have omitted from the calculation for some reason.

This file lists all A1-5 x B1-2 x C1-5 permutations. The model you required I use to generate the set they are all possible - the commented out ones are just presenting a FATAL error of some kind so the current testing mechanism cannot handle the test result. In order to use your model I have had to drop testing of the edge cases that made the earlier permutation matrix sparse.

Reply:
> The extra 5 lines are the (A5 x B2 x C1-5) set you have omitted from the calculation for some reason.

Aha, I think I have discovered the source of the disagreement: You think that A5 (i.e. "omitted tls-cert contents") is compatible with B2 (i.e. tls-cert contains a "merged" certificate and key). I disagree because "omitted" means "not present" to me, and B2 certainly requires `tsl-cert` contents.

I think this problem can be addressed if, when defining A and B groups as a whole, instead of using `tls-cert` and `tls-key` option names, we use the thing the group is focusing on:

* For group A, that thing is "certificate(s)". 
* For group B, that thing is "certificate key".

Here is the corresponding sketch:

```
# The following port configurations check all 50 possible permutations of these
# configs. Among those, the negative ones (unsupported and invalid) are
# currently commented out. TODO: Check each negative test case in isolation.
#
#  A) certificate(s):
#    1) leaf only         tls-cert=./leaf-rsa.*
#    2) full chain        tls-cert=./leaf-mid-root-rsa.*
#    3) root CA           tls-cert=./root-rsa.*
#    4) incomplete chain  tls-cert=./mid-rsa.*
#    5) omitted (no tls-cert for B1 or key-only tls-cert for B2)
#
#  B) signing certificate key:
#    1) in a dedicate file          tls-key=./*.pkey
#    2) in a tls-cert file          tls-cert=./*.pem
```

In the above sketch, I also polished file names in the A1-4 group to look more consistent by always enumerating the stored certificates. If you reject this part of the polishing for some reason, please at least remove `key` from `leaf-key-rsa.*` -- the key location is determined by `B`.

The above change will make `A5 x B1-2 x C1-5` combinations possible, adding the 5 previously "missing" cases: `A1-5 x B1-2 x C1-5` is `5 * 2 * 5` is `50`.

Reply:
> I think this problem can be addressed if, when defining A and B groups as a whole, instead of using `tls-cert` and `tls-key` option names, we use the thing the group is focusing on:
> 
>     * For group A, that thing is "certificate(s)".
> 
>     * For group B, that thing is "certificate key".
> 

That is what I have been focusing on from the start for these two groups. If this text seems clearer to you I'm okay with it and will add shortly.

Reply:
Great, thank you.

Reply:
done in 56b41ce